platform/pulp/platform/models/progress.py

```diff
@@ -0,0 +1,195 @@
+"""
+Django models related to Progress Reporting
+"""
+from gettext import gettext as _
+import logging
+
+from django.contrib.contenttypes import fields
+from django.db import models
+
+from pulp.platform.models import Model, Task
+
+
+__all__ = [
+    'ProgressBar',
+    'ProgressReportingBase',
+    'ProgressSpinner',
+]
+
+
+_logger = logging.getLogger(__name__)
+
+
+class ProgressReportingBase(Model):
+    """
+    An abstract base model for all progress reporting.
+
+    All progress reports have a message, state, and are related to a Task.
+
+    Fields:
+
+    :cvar message: A short message for the progress update, typically shown to the user. (required)
+    :type message: models.TextField
+
+    :cvar state: The state of the progress update. Defaults to `WAITING`. This field uses a limited
+                 set of choices of field states. See `STATES` for possible states.
+    :type state: models.TextField
+
+    Relations:
+
+    :cvar task: The task associated with this progress report. If left unset when save() is called
+                it will be set to the current task_id.
+    :type task: models.ForeignKey
+    """
+    WAITING = 'waiting'
+    SKIPPED = 'skipped'
+    RUNNING = 'running'
+    FINISHED = 'finished'
+    ERROR = 'error'
+    CANCELED = 'canceled'
+    STATES = (
+        (WAITING, 'Waiting'),
+        (SKIPPED, 'Skipped'),
+        (RUNNING, 'Running'),
+        (FINISHED, 'Finished'),
+        (ERROR, 'Error'),
+        (CANCELED, 'Canceled')
+    )
+    message = models.TextField()
+    state = models.TextField(choices=STATES, default=WAITING)
+    task = fields.GenericRelation(Task)
+
+    class Meta:
+        abstract = True
+
+    def call_and_observe(self, func, args, kwargs):
```

Comment:
This looks easy to use. Another helpful way to use this would be as a context manager. I think the use would be something like:

```
report = ProgressSpinner(...)
with report:
    do_stuff()
```

Then you could just put the code in the ``try`` block below in an ``__enter__`` method, and everything else in the ``__exit__`` method.

The ``__enter__`` method could optionally return something to track a loop, similar to what you've done below with  ``ProgressBar``. A from-the-hip rough-draft of use:

```
things = get_me_some_things()
report = ProgressBar(max=len(things))
with report as loop_watcher:
    for thing in loop_watcher(things):
        do_stuff(thing)
```

or some might prefer this:

```
things = get_me_some_things()
report = ProgressBar(max=len(things))
with report as loop_watcher:
    for thing in things:
        do_stuff(thing)
        loop_watcher.item_done()
```